storage/writer.go

```diff
@@ -84,30 +94,301 @@ type Writer struct {
 
 	mu  sync.Mutex
 	err error
+
+	stream storagepb.Storage_WriteObjectClient
 }
 
-func (w *Writer) open() error {
-	attrs := w.ObjectAttrs
-	// Check the developer didn't change the object Name (this is unfortunate, but
-	// we don't want to store an object under the wrong name).
-	if attrs.Name != w.o.object {
-		return fmt.Errorf("storage: Writer.Name %q does not match object name %q", attrs.Name, w.o.object)
+// openGRPC initializes a pipe for the user to write data to, and a routine to
+// read from that pipe and upload the data to GCS via gRPC.
+//
+// This is an experimental API and not intended for public use.
+func (w *Writer) openGRPC() error {
+	if err := w.validateWriteAttrs(); err != nil {
+		return err
 	}
-	if !utf8.ValidString(attrs.Name) {
-		return fmt.Errorf("storage: object name %q is not valid UTF-8", attrs.Name)
+
+	pr, pw := io.Pipe()
+	w.pw = pw
+	w.opened = true
+
+	go w.monitorCancel()
+
+	// TODO: Handle this case later.
+	bufSize := w.ChunkSize
+	if w.ChunkSize == 0 {
+		bufSize = maxPerMessageWriteSize
 	}
-	if attrs.KMSKeyName != "" && w.o.encryptionKey != nil {
-		return errors.New("storage: cannot use KMSKeyName with a customer-supplied encryption key")
+
+	var offset int64
+	var upid string
+
+	go func() {
+		defer close(w.donec)
+
+		// Loop until there is an error or the Object has been finalized.
+		for {
+			// Initiliaze client buffer with ChunkSize.
+			buf := make([]byte, bufSize)
```

Comment:
This buffer should be able to be reused between chunks, right? If you re-create it in each loop iteration this will be very costly.

Reply:
Good point! I will just truncate the slice, which should keep the underlying array memory allocation. Correct me if I'm wrong.

Reply:
I wasn't sure if this was the correct approach, but this is what `bytes.Buffer.Reset()` does and what was recommended on the interwebs :)

Reply:
That seems to cause a problem, namely `Read` won't resize the buffer (obviously), so it just doesn't read anything and we get a `nil` panic at a later stage...

Reply:
Ok I think I've fixed it. Just keep reusing the same buffer to `read` but provide a sub-slice of `recvd` bytes to the `uploadBuffer` call.

Reply:
Discussed offline. Done!